src/main/java/org/gaul/s3proxy/azureblob/AzureBlobStore.java

```diff
@@ -571,65 +584,436 @@ public void setBlobAccess(String container, String key, BlobAccess access) {
+                        "Conflict during commit: " + bse.getMessage(), bse);
+            } else if (bse.getStatusCode() == 412) {
+                translateAndRethrowException(bse, mpu.containerName(), targetBlobName);
+            }
+            throw bse;
         }
-        var blockBlobItem = client.commitBlockList(blocks.build(),
-                /*overwrite=*/ true);
-        return blockBlobItem.getETag();
     }
 
     @Override
     public MultipartPart uploadMultipartPart(MultipartUpload mpu,
             int partNumber, Payload payload) {
-        var client = blobServiceClient
-                .getBlobContainerClient(mpu.containerName())
-                .getBlobClient(mpu.blobName())
-                .getBlockBlobClient();
-        var blockId = makeBlockId(partNumber);
-        var length = payload.getContentMetadata().getContentLength();
-        try (var is = payload.openStream()) {
-            client.stageBlock(blockId, is, length);
+
+        if (partNumber < 1 || partNumber > 10_000) {
+            throw new IllegalArgumentException(
+                    "Part number must be between 1 and 10,000, got: " + partNumber);
+        }
+
+        Long contentLength = payload.getContentMetadata().getContentLength();
+        if (contentLength == null) {
+            throw new IllegalArgumentException("Content-Length is required");
+        }
+        if (contentLength < 0) {
+            throw new IllegalArgumentException(
+                    "Content-Length must be non-negative, got: " + contentLength);
+        }
+
+        if (contentLength > getMaximumMultipartPartSize()) {
+            throw new IllegalArgumentException(
+                    "Part size exceeds maximum of " + getMaximumMultipartPartSize() +
+                    " bytes: " + contentLength);
+        }
+
+
+        String uploadKey = mpu.id();
+        String nonce = uploadKey.substring(STUB_BLOB_PREFIX.length());
+        String blockId = makeBlockId(nonce, partNumber);
+        var asyncClient = createNonRetryingBlockBlobAsyncClient(
+                mpu.containerName(), mpu.blobName());
+
+        byte[] md5Hash;
+        try {
+            var is = payload.openStream();
+            var providedMd5 = payload.getContentMetadata().getContentMD5AsHashCode();
+            var md = MessageDigest.getInstance("MD5");
+
+            Flux<ByteBuffer> body = Flux.generate(
+                () -> 0L,
+                (position, sink) -> {
+                    try {
+                        if (position >= contentLength) {
+                            sink.complete();
+                            return position;
+                        }
+                        int chunkSize = (int) Math.min(4 * 1024 * 1024L,
+                                contentLength - position);
+                        ByteBuffer buffer = ByteBuffer.allocate(chunkSize);
+                        int read = is.read(buffer.array(), 0, chunkSize);
```

Comment:
I believe that a call to `InputStream.read` may return fewer bytes than requested yet more may come from a socket in a subsequent call.  So you really want a loop:

```java
for (int i = 0; i < chunkSize; ) {
    int read = is.read(buffer.array(), i, buffer.remaining());
    if (read == -1) { ... }
    i += read;
    ...
}
```